app/mempool/lane.go

```diff
@@ -0,0 +1,279 @@
+		if transactionLimit.IsExceededBy(txInfo.BlockSpace) {
+			l.logger.Info(
+				"failed to select tx for lane; tx exceeds limit",
+				"tx_hash", txInfo.Hash,
+				"lane", l.name,
+			)
+
+			txsToRemove = append(txsToRemove, tx)
+			continue
+		}
+
+		// Add the transaction to the proposal.
+		// TODO: check if the transaction cannot be added here, it should also cannot be added afterward.
+		if err := proposal.Add(txInfo); err != nil {
+			l.logger.Info(
+				"failed to add tx to proposal",
+				"lane", l.name,
+				"tx_hash", txInfo.Hash,
+				"err", err,
+			)
+
+			break
+		}
+
+		blockUsed = blockUsed.Add(txInfo.BlockSpace)
+	}
+
+	return
+}
+
+// FillProposalBy fills the proposal with transactions from the lane mempool with the given iterator and limit.
+// It returns the total size and gas of the transactions added to the proposal.
+// It also returns a list of transactions that were removed from the lane mempool.
+func (l *Lane) FillProposalBy(
+	proposal *Proposal,
+	iterator sdkmempool.Iterator,
+	laneLimit BlockSpace,
+) (blockUsed BlockSpace, txsToRemove []sdk.Tx) {
+	// get the transaction limit for the lane.
+	transactionLimit := proposal.GetLimit(l.maxTransactionSpace)
+
+	// Select all transactions in the mempool that are valid and not already in the
+	// partial proposal.
+	for ; iterator != nil; iterator = iterator.Next() {
+		// If the total size used or total gas used exceeds the limit, we break and do not attempt to include more txs.
+		// We can tolerate a few bytes/gas over the limit, since we limit the size of each transaction.
+		if laneLimit.IsReachedBy(blockUsed) {
+			break
+		}
+
+		tx := iterator.Tx()
+		txInfo, err := l.GetTxInfo(tx)
+		if err != nil {
+			l.logger.Info("failed to get hash of tx", "err", err)
+
+			txsToRemove = append(txsToRemove, tx)
+			continue
+		}
+
+		// if the transaction is exceed the limit, we remove it from the lane mempool.
+		if transactionLimit.IsExceededBy(txInfo.BlockSpace) {
+			l.logger.Info(
+				"failed to select tx for lane; tx exceeds limit",
+				"tx_hash", txInfo.Hash,
+				"lane", l.name,
+			)
+
+			txsToRemove = append(txsToRemove, tx)
+			continue
+		}
+
+		// Add the transaction to the proposal.
+		if err := proposal.Add(txInfo); err != nil {
+			l.logger.Info(
+				"failed to add tx to proposal",
+				"lane", l.name,
+				"tx_hash", txInfo.Hash,
+				"err", err,
+			)
+
+			break
+		}
+
+		// Update the total size and gas.
+		blockUsed = blockUsed.Add(txInfo.BlockSpace)
+	}
+
+	return
+}
+
+// GetTxInfo returns various information about the transaction that
+// belongs to the lane including its priority, signer's, sequence number,
+// size and more.
+func (l *Lane) GetTxInfo(tx sdk.Tx) (TxWithInfo, error) {
+	txBytes, err := l.txEncoder(tx)
+	if err != nil {
+		return TxWithInfo{}, fmt.Errorf("failed to encode transaction: %w", err)
+	}
+
+	// TODO: Add an adapter to lanes so that this can be flexible to support EVM, etc.
```

Comment:
is this going to be done in this pr? if so just bumping

Reply:
deleted